---
title: "1_data_cleaning"
author: "Joy_Fu"
date: "2023-03-31"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
rm(list = ls())
pacman::p_load(tidyverse, PheWAS)
raw_data_path = "/Users/Mingzhou/Desktop/Projects/AD-LOE-genetics/data/"
output_path = "/Users/Mingzhou/Desktop/Projects/AD-LOE-genetics/output/"
# Source in useful functions
source("functions.R")
```

# Part 1. Raw EHR data preprocessing
## 1. Clean demographic info
```{r}
phe_table_extract_date = "2023-03-20"
# Read in data
atlas_data_path = "/Users/Mingzhou/Desktop/Projects/Genetic-prediction-dementia/data/"
demo_raw = read_csv(file = paste0(atlas_data_path, 
                                  "atlas/pheno/raw/atlas_demo_03202023.csv"),
                    col_names = FALSE, na = c("NULL")) 
dim(demo_raw) # dim = (436258,8)
names(demo_raw) = c("PatientID", "UniqueSampleId", "Sex", "Ethnicity", "FirstRace", 
                    "BirthDate", "DeathDate", "PatientLivingStatus")
# Clean demographic info
ddr_demo_short = demo_raw %>% 
  filter(Sex != "X" & Sex != "Unknown") %>% 
  dplyr::select(PatientID, UniqueSampleId, BirthDate, DeathDate) %>% 
  unique() %>% filter(!is.na(UniqueSampleId))
dim(ddr_demo_short) # dim = (65479,4)
```

## 2. Clean encounter info
```{r}
# Read in data
enc_raw = read_csv(file = paste0(atlas_data_path, 
                                 "atlas/pheno/raw/atlas_encounter_03202023.csv"),
                   col_names = FALSE, na = c("NULL")) 
dim(enc_raw) # dim = (30399114,5)
names(enc_raw) = c("PatientID", "UniqueSampleId", "Sex", "DiagnosisCode",
                   "StartDate")
enc_demo = enc_raw %>% 
  mutate(StartDate = as.Date(StartDate)) %>% 
  inner_join(ddr_demo_short) %>% 
  mutate(age_diagnosis = as.numeric(StartDate - BirthDate)/365.25) %>% 
  filter(!is.na(age_diagnosis)) %>% 
  arrange(PatientID, StartDate) %>% 
  mutate(valid_record = case_when(
    is.na(DeathDate) & StartDate >= BirthDate ~ 1,
    !is.na(DeathDate) & StartDate >= BirthDate & StartDate <= DeathDate ~ 1
  )) %>% filter(valid_record == 1) %>% select(-valid_record)
dim(enc_demo) # dim = (29628667,8)
```

## 3. Add genetic ancestry and EHR features
```{r}
load(file = paste0(atlas_data_path, "atlas/geno/gen_ancestry.rda"))
enc_demo_60 = enc_demo %>% 
  inner_join(gen_ancestry) %>% filter(age_diagnosis >= 60) 
dim(enc_demo_60) # dim = (13934109,9)
patient_ehr = enc_demo_60 %>% 
  mutate(ICD_3digit = substr(DiagnosisCode, 1, 3)) %>% 
  group_by(PatientID) %>% 
  mutate(age_last_visit = as.numeric(max(StartDate) -BirthDate)/365.25) %>% 
  mutate(record_length = as.numeric(max(StartDate) - min(StartDate))/365.25) %>% 
  mutate(n_diagnosis = length(unique(ICD_3digit))) %>% 
  mutate(n_encounter = length(unique(StartDate))) %>% ungroup() %>% 
  select(PatientID, UniqueSampleId, Sex, gen_ancestry, age_last_visit,
         record_length, n_diagnosis, n_encounter) %>% unique() %>% 
  # filter those age at last visit 60-90 and at least 2 visits
  filter(age_last_visit >= 60 & age_last_visit < 90) %>% 
  filter(n_encounter >= 2)
dim(patient_ehr) # dim = (25018,8)
```

## 4. Phecode w/ PheWAS package 
```{r}
enc_demo_filtered = enc_demo %>% 
  filter(UniqueSampleId %in% patient_ehr$UniqueSampleId) %>% 
  group_by(UniqueSampleId, DiagnosisCode) %>% 
  arrange(StartDate) %>% slice_head(n = 1) %>% ungroup()
dim(enc_demo_filtered) # dim = (2818597,8)
# Map to phecodes
enc_phecode_map = enc_demo_filtered %>% 
  mutate(vocabulary_id = "ICD10CM") %>% 
  left_join(PheWAS::phecode_map, by = c("vocabulary_id" = "vocabulary_id",
                                        "DiagnosisCode" = "code")) %>% unique()
dim(enc_phecode_map) # dim = (2885262,10)
# Collapse to phecode-level first diagnosis
enc_phecode_map_first = enc_phecode_map %>% 
  group_by(UniqueSampleId, phecode) %>% 
  arrange(StartDate) %>% slice_head(n = 1) %>% ungroup() %>% unique() %>% 
  select(-DiagnosisCode)
dim(enc_phecode_map_first) # dim = (1766708,8)
# Save the phenotype data table
save(enc_phecode_map_first, 
     file = paste0(raw_data_path, "atlas/pheno/enc_phecode_map_first.rda"))
```


# Part 2. Sample selection
## 1. Define epilepsy
```{r}
# Select epilepsy case and controls
epi_phecode = c("345", "345.1", "345.11", "345.12", "345.3")
epi_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% epi_phecode) %>% 
  filter(age_diagnosis >= 60) %>% group_by(UniqueSampleId) %>% 
  mutate(epi_age_diagnosis = min(age_diagnosis),
         EpilepsyDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, EpilepsyDate, epi_age_diagnosis) %>% 
  unique() %>% left_join(patient_ehr) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90) %>% mutate(epilepsy = 1)
dim(epi_case_info) # dim = (887,11)
```

```{r}
# exclude those with confounding phenotype (exclusion criteria by PheWAS)
epi_exclude_phecode = PheWAS::phecode_exclude %>% 
  filter(code %in% epi_phecode) %>% pull(exclusion_criteria) %>% unique()
epi_control_exclude_Id = enc_phecode_map_first %>% 
  filter(UniqueSampleId %!in% epi_case_info$UniqueSampleId) %>% 
  filter(phecode %in% epi_exclude_phecode) %>% pull(UniqueSampleId) %>% unique()
# N = 5112
# control info
epi_control_info = patient_ehr %>% 
  filter(UniqueSampleId %!in% epi_case_info$UniqueSampleId) %>% 
  filter(UniqueSampleId %!in% epi_control_exclude_Id) %>% 
  mutate(epilepsy = 0, EpilepsyDate = NA, epi_age_diagnosis = NA) %>% 
  select(names(epi_case_info)) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90)
dim(epi_control_info) # dim = (19019,11)
```

```{r}
# combine cases and controls
epi_patients_full = rbind(epi_case_info, epi_control_info) %>% as.data.frame()
dim(epi_patients_full) # dim = (19906,11)
```

## 2. Define dementia
```{r}
dem_phecode = c("290.1", "290.11", "290.12", "290.16")
dem_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% dem_phecode) %>% group_by(UniqueSampleId) %>% 
  mutate(dem_age_diagnosis = min(age_diagnosis),
         DementiaDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, DementiaDate, dem_age_diagnosis) %>% 
  unique() %>% left_join(patient_ehr) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90) %>% mutate(dementia = 1)
dim(dem_case_info) # dim = (1157,11)
```

```{r}
# exclude those with confounding phenotype (exclusion criteria by PheWAS)
dem_exclude_phecode = PheWAS::phecode_exclude %>% 
  filter(code %in% dem_phecode) %>% pull(exclusion_criteria) %>% unique()
dem_control_exclude_Id = enc_phecode_map_first %>% 
  filter(UniqueSampleId %!in% dem_case_info$UniqueSampleId) %>% 
  filter(phecode %in% dem_exclude_phecode) %>% pull(UniqueSampleId) %>% unique()
# N = 5357
# control info
dem_control_info = patient_ehr %>% 
  filter(UniqueSampleId %!in% dem_case_info$UniqueSampleId) %>% 
  filter(UniqueSampleId %!in% dem_control_exclude_Id) %>% 
  mutate(dementia = 0, DementiaDate = NA, dem_age_diagnosis = NA) %>% 
  select(names(dem_case_info)) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90)
dim(dem_control_info) # dim = (18504,11)
```

```{r}
# combine cases and controls
dem_patients_full = rbind(dem_case_info, dem_control_info) %>% as.data.frame()
dim(dem_patients_full) # dim = (19661,11)
```

## 3. Define AD
```{r}
ad_phecode = c("290.11")
ad_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% ad_phecode) %>% group_by(UniqueSampleId) %>% 
  mutate(ad_age_diagnosis = min(age_diagnosis),
         ADDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, ADDate, ad_age_diagnosis) %>% 
  unique() %>% left_join(patient_ehr) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90) %>% mutate(AD = 1)
dim(ad_case_info) # dim = (473,11)
```

```{r}
# exclude those with confounding phenotype (exclusion criteria by PheWAS)
ad_exclude_phecode = PheWAS::phecode_exclude %>% 
  filter(code %in% ad_phecode) %>% pull(exclusion_criteria) %>% unique()
ad_control_exclude_Id = enc_phecode_map_first %>% 
  filter(UniqueSampleId %!in% ad_case_info$UniqueSampleId) %>% 
  filter(phecode %in% ad_exclude_phecode) %>% pull(UniqueSampleId) %>% unique()
# N = 6041
# control info
ad_control_info = patient_ehr %>% 
  filter(UniqueSampleId %!in% ad_case_info$UniqueSampleId) %>% 
  filter(UniqueSampleId %!in% ad_control_exclude_Id) %>% 
  mutate(AD = 0, ADDate = NA, ad_age_diagnosis = NA) %>% 
  select(names(ad_case_info)) %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90)
dim(ad_control_info) # dim = (18504,11)
```

```{r}
# combine cases and controls
ad_patients_full = rbind(ad_case_info, ad_control_info) %>% as.data.frame()
dim(ad_patients_full) # dim = (18977,11)
```

## 4. Left join find overlap
```{r}
patient_eligible = epi_patients_full %>% full_join(dem_patients_full) %>% 
  full_join(ad_patients_full) %>% 
  filter(record_length > 0 & n_encounter >= 2)
dim(patient_eligible) # dim = (23017,17)
save(patient_eligible,
     file = paste0(raw_data_path, "atlas/pheno/patient_eligible.rda"))
```

## 5. Also add other covariates
### 1) HTN
```{r}
htn_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% c("401.1")) %>% group_by(UniqueSampleId) %>% 
  mutate(htn_age_diagnosis = min(age_diagnosis),
         HTNDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, HTNDate, htn_age_diagnosis) %>% unique()
dim(htn_case_info) # dim = (17204,4)
```

### 2) Diabetes
```{r}
diab_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% c("250.2")) %>% group_by(UniqueSampleId) %>% 
  mutate(diab_age_diagnosis = min(age_diagnosis),
         DiabetesDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, DiabetesDate, diab_age_diagnosis) %>% unique()
dim(diab_case_info) # dim = (6649,4)
```

### 3) Stroke
```{r}
stroke_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% c("433")) %>% group_by(UniqueSampleId) %>% 
  mutate(stroke_age_diagnosis = min(age_diagnosis),
         StrokeDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, StrokeDate, stroke_age_diagnosis) %>% unique()
dim(stroke_case_info) # dim = (558,4)
```

### 4) Hyperlipid
```{r}
hyperlipid_case_info = enc_phecode_map_first %>% 
  filter(phecode %in% c("272.1")) %>% group_by(UniqueSampleId) %>% 
  mutate(hyperlipid_age_diagnosis = min(age_diagnosis),
         HyperlipidDate = min(StartDate)) %>% 
  select(PatientID, UniqueSampleId, HyperlipidDate, hyperlipid_age_diagnosis) %>% unique()
dim(hyperlipid_case_info) # dim = (17106,4)
```


## 5. Combine all
```{r}
# Add PCs and APOE
geno_path = "/Users/Mingzhou/Desktop/Projects/Genetic-prediction-dementia/data/atlas/geno/"
# Add genetic info
load(paste0(geno_path, "apoe.rda"))
load(paste0(geno_path, "pcs/eur_pcs.rda"))
apoe$UniqueSampleId = as.numeric(apoe$UniqueSampleId)
eur_pcs$UniqueSampleId = as.numeric(eur_pcs$UniqueSampleId)

patient_final = patient_eligible %>% 
  left_join(htn_case_info) %>% left_join(diab_case_info) %>% 
  left_join(stroke_case_info) %>% left_join(hyperlipid_case_info) %>% 
  mutate(female = if_else(Sex == "Female", 1, 0)) %>% 
  mutate(enc_per_yr = n_encounter/record_length) %>% 
  select(UniqueSampleId, female, gen_ancestry, age_last_visit, 
         record_length, enc_per_yr, n_diagnosis, n_encounter,
         epilepsy, EpilepsyDate, epi_age_diagnosis,
         AD, ADDate, ad_age_diagnosis,
         dementia, DementiaDate, dem_age_diagnosis,
         HTNDate, htn_age_diagnosis,
         DiabetesDate, diab_age_diagnosis,
         StrokeDate, stroke_age_diagnosis,
         HyperlipidDate, hyperlipid_age_diagnosis)
dim(patient_final) # dim = (23017,25)
save(patient_final, file = paste0(raw_data_path, 
                                  "modeling/patient_final.rda"))

# Add AD PRSs
load(file = paste0(raw_data_path, "prs/mod/AD_PRS_final.rda"))
patient_final_prs = patient_final %>% 
  left_join(AD_PRS_final)
dim(patient_final_prs) # dim = (23017,27)

patient_final_EUR = patient_final_prs %>% 
  left_join(apoe) %>% left_join(eur_pcs) %>% 
  filter(gen_ancestry == "EUR") %>% 
  filter(age_last_visit >= 60 & age_last_visit < 90)
dim(patient_final_EUR) # dim = (17306,50)
save(patient_final_EUR, file = paste0(raw_data_path, 
                                      "modeling/EUR/patient_final_EUR.rda"))
```


# Part 3. FUMA results cleaning
```{r}
rm(list = ls())
pacman::p_load(tidyverse, GenomicRanges, rtracklayer, bigsnpr, bigreadr, 
               data.table, ieugwasr, MRutils)
raw_data_path = "/Users/Mingzhou/Desktop/Projects/AD-LOE-genetics/data/"
output_path = "/Users/Mingzhou/Desktop/Projects/AD-LOE-genetics/output/"
FUMA_data_path = paste0(raw_data_path, "FUMA/")
# Source in useful functions
source("functions.R")
```

## 1. Find matched SNPs
### 1) AD phenotype
```{r}
candidate_snps_AD = read.table(file = paste0(FUMA_data_path, "FUMA_AD_new/snps.txt"), 
                               header = T, sep = "\t", fill = T)
candidate_snps_short_AD = candidate_snps_AD %>% 
  dplyr::rename("a0" = "non_effect_allele", "a1" = "effect_allele") %>% 
  filter(posMapFilt == 1 | eqtlMapFilt == 1 | ciMapFilt == 1) %>% 
  select(rsID, chr, pos, a0, a1, gwasP, CADD, r2, IndSigSNP, GenomicLocus) %>% 
  mutate(phenotype = "AD")
dim(candidate_snps_short_AD) # dim = (2089,11)
```

### 2) Epilepsy phenotype
```{r}
candidate_snps_epilepsy = read.table(file = paste0(FUMA_data_path, "FUMA_general_epilepsy_new/snps.txt"), 
                                     header = T, sep = "\t", fill = T)
candidate_snps_short_epilepsy = candidate_snps_epilepsy %>% 
  dplyr::rename("a0" = "non_effect_allele", "a1" = "effect_allele") %>% 
  filter(posMapFilt == 1 | eqtlMapFilt == 1 | ciMapFilt == 1) %>% 
  select(rsID, chr, pos, a0, a1, gwasP, CADD, r2, IndSigSNP, GenomicLocus) %>% 
  mutate(phenotype = "epilepsy")
dim(candidate_snps_short_epilepsy) # dim = (678,11)
```

```{r}
# Check if there are overlap (none)
intersect(candidate_snps_short_AD$rsID, candidate_snps_short_epilepsy$rsID)
intersect(candidate_snps_short_AD$IndSigSNP, candidate_snps_short_epilepsy$IndSigSNP)
# Combine all SNPs
candidate_snps_combined = rbind(candidate_snps_short_AD, candidate_snps_short_epilepsy)
```

## 2. Cleaning (liftover + annotation)
```{r message=FALSE, warning=FALSE}
setwd("~/Desktop/Projects/AD_genetic_pred/code")
chain_file = import.chain("hg19ToHg38.over.chain")
candidate_hg19 = candidate_snps_combined %>% 
  mutate(No = seq(1,nrow(candidate_snps_combined)))
# Prepare Genomic Ranges
hg19.gr = GRanges(
  seqname = paste("chr", candidate_hg19$chr, sep = ""),
  ranges = IRanges(start = as.numeric(candidate_hg19$pos), 
                   end = as.numeric(candidate_hg19$pos)),
  snp.name = candidate_hg19$rsID)
candidate_hg38 = as.data.frame(liftOver(hg19.gr, chain_file)) %>% 
  mutate(chr = as.integer(substring(seqnames, 4, last = 1000000L))) %>% 
  dplyr::select(group, chr, start) %>% unique() %>% drop_na()
candidate_snps_liftover = candidate_hg19 %>% 
  left_join(candidate_hg38, 
            by = c("No" = "group", "chr" = "chr")) %>% select(-pos) %>% 
  dplyr::rename("pos" = "start") %>% 
  filter(!is.na(chr) & !is.na(pos)) %>% unique() 
dim(candidate_snps_liftover) # dim = 2756,12
# 3. Extract and clean ATLAS data -- Server step
snp_to_extract = candidate_snps_liftover %>% 
  mutate(chr_pos_name1 = paste0("chr", chr, ":", pos, ":", a0, ":", a1),
         chr_pos_name2 = paste0("chr", chr, ":", pos, ":", a1, ":", a0))
name1_out = snp_to_extract %>% pull(chr_pos_name1) %>% unique()
name2_out = snp_to_extract %>% pull(chr_pos_name2) %>% unique()
name_out_full = c(name1_out, name2_out) %>% unique()
length(name_out_full) # length = 5512
# Write output for snp extraction on server
write.table(name_out_full,
            file = paste0(raw_data_path, "sumstats/AD_epilepsy_wdup.txt"),
            sep = "\t", quote = F, row.names = F, col.names = F)
```


```{r}
# Read in the QCed genotype and sumstats files + preprocess the bed file (only need to do once for each data set)
geno_file = paste0(raw_data_path, 'atlas/geno/AD.epilepsy.combine')
snp_readBed(paste0(geno_file, '.bed'))
obj.bigSNP = snp_attach(paste0(geno_file, '.rds'))
# extract the SNP information from the genotype
map = obj.bigSNP$map[-3]
names(map) = c("chr", "uniqueID", "pos", "a1", "a0")
# Reverse beta if necessary by performing SNP matching
rsid_map_beta_align = snp_to_extract %>% 
  select(rsID, chr, pos, a1, a0, gwasP, CADD) %>% 
  dplyr::rename("beta" = "CADD") %>% unique()
info_snp = snp_match(rsid_map_beta_align, map) # 1958 variants have been matched
```

## 2. Select mapped SNPs
```{r}
mapped_snp = info_snp %>% select(rsID, uniqueID) %>% 
  left_join(snp_to_extract) %>% 
  group_by(IndSigSNP) %>% arrange(desc(CADD)) %>% 
  mutate(rank_CADD = row_number()) %>% ungroup() %>% 
  filter(rank_CADD == 1 | rsID == IndSigSNP)
save(mapped_snp, file = paste0(raw_data_path, "modeling/EUR/mapped_snp.rda"))
# Check how many mapped SNPs are IndSigSNP (N = 97)
ttt = mapped_snp %>% group_by(IndSigSNP) %>% summarise(n = n())
table(ttt$n)
# Extract SNPs
mapped_final_snp = mapped_snp %>% filter(rank_CADD == 1) %>% 
  pull(uniqueID) %>% unique()
write.table(mapped_final_snp,
            file = paste0(raw_data_path, "sumstats/mapped_final_snp.txt"),
            sep = "\t", quote = F, row.names = F, col.names = F)
```

```{r}
# Get genotyped data
genotype = obj.bigSNP$genotypes
genotype.targ = snp_fastImputeSimple(obj.bigSNP$genotypes, method = 'mean2') 
genotype_df = as.matrix(genotype.targ[])[,] %>% as.data.frame()
colnames(genotype_df) = map$uniqueID
dim(genotype_df) # 34463,2272
rownames(genotype_df) = obj.bigSNP$fam$sample.ID
```


```{r}
# Perform final matching using selected CADD-1st (& CADD >= 12.37)
final2map = mapped_snp %>% filter(rank_CADD == 1) %>% 
  select(rsID, chr, pos, a1, a0, gwasP, phenotype, CADD, GenomicLocus) %>% 
  dplyr::rename("beta" = "CADD") %>% unique()
final_map = snp_match(final2map, map) # 162 variants have been matched
```

## 3. Finalize genome dataset
```{r}
genotype_df_map = genotype_df %>% select(all_of(final_map$uniqueID))
dim(genotype_df_map) # 34463,162
save(genotype_df_map, file = paste0(raw_data_path, 
                                    "modeling/EUR/genotype_df_map.rda"))
# Need to exclude some columns (w/ large amount of missing data)
ttt = colSums(is.na(genotype_df_map))
length(ttt[ttt > 10000]) # N = 0 snps that have missing value > 10000
# Remove zero variance
zero_freq_col = zeroVar(genotype_df_map)
length(zero_freq_col) # N = 0
```

```{r}
load(file = paste0(raw_data_path, "modeling/EUR/patient_final_EUR.rda"))

# Full genome data + epilepsy + demo
freeze60k_geno_freq = genotype_df_map %>% 
  rownames_to_column(var = "UniqueSampleId") %>% 
  mutate(UniqueSampleId = as.numeric(UniqueSampleId)) %>% 
  # add phenotype info
  inner_join(patient_final_EUR) %>% column_to_rownames(var = "UniqueSampleId")
dim(freeze60k_geno_freq) # dim = c(17306, 211)
save(freeze60k_geno_freq, file = paste0(raw_data_path, 
                                       "modeling/EUR/freeze60k_geno_freq.rda"))
```

